### DISCUSSION TITLE
remove need to have unique index on relations with more than 1 field and move to index (Not unique)

### DISCUSSION THREAD

**Discussion by marco-mastella** [prisma/prisma#22650] [Q&A]
## Problem
let's assume we have this schema
```
model Company{
    id                      String    @id @default(uuid())
    companyName             String?   @default("")
}

model Registry{
    id                      String    @id @default(uuid())
    idCompany               String?   @default("")
    registryName            String?   @default("")
    company                 Company?  @relation(fields: [idCompany], references: [id])
}

model Activity{
    id                      String    @id @default(uuid())
    idCompany               String?   @default("")
    idRegistry              String?   @default("")
    description             String?   @default("")
    registry                Registry? @relation(fields: [idCompany, idRegistry], references: [id, idCompany])
}
```

Error parsing attribute "@relation": The argument references must refer to a unique criterion in the related model. Consider adding an @@unique([id, idCompany]) attribute to the model Registry.

## Suggested solution

instead of unique index i would suggest to create an index (not unique)

**Comment by SevInf** [comment]
Without unique index, 1:1 relation can not be guaranteed. In you schema, as per definitions you've postded here it is possible that multiple Acvtivites with the same `(idCompany, idRegistry)` combination will be defined making it 1:m relation instead. Unique index will prevent that.

**Reply by marco-mastella** [reply]
that's what I want to achieve, have multiple activities with the same idCompany, idRegistry.

**Reply by SevInf** [reply]
In that case, `Registry.id` is a primary key and already identifies the registry activity belongs to, there is no need to have `idCompany` as a part of relation.

**Reply by marco-mastella** [reply]
i understand your point, i tried to create an example avoiding to provide too much details to focus on main point (the needs of a unique index). 
Let's assume `Registry.id` is not an unique index (and also could not be an unique index `registry.id registry.idCompany`) and i need to create a relation between the 2 models, accepting that could be a 1:N relation. How can i proceed considering that the relation is based on the 2 fields `idCompany` and `idRegistry`?
I don't see (from a DB prospective, on ORM prospective i'm trying to understand) why it is mandatory to have an unique index. Also based on your latest features on relation `relationLoadStrategy: 'query'`

**Reply by SevInf** [reply]
Without unique index on the other side, it is m:n relation, not m:1. Basically, assuming `id` is not unique in that case, it is possible to create two registries with the same `(id, idCompany)` combination. So, that part in your schema where each activity belongs to a single registry might not be true without a unique index.

If m:n relation between registry and activity is what you want, [explicit many-to-many](https://www.prisma.io/docs/orm/prisma-schema/data-model/relations/many-to-many-relations#explicit-many-to-many-relations) is how you can model that with Prisma.

**Reply by marco-mastella** [reply]
the n:m relation is not the case i'm trying to analyse. I need to create a clearer example in order to give you a better understanding of the problem.

**Reply by marco-mastella** [reply]
Please consider the following schema
```
model ModelA{
  id                        String                @id @default(uuid())
  preferredModelBId String?               @default("")
  modelB          ModelB[]    @relation("relationList")
  preferredModelB   ModelB?     @relation("relationPreferred")

  @@unique([preferredModelBId ])
}

model ModelB{
  id                    String                @id @default(uuid())
  idModelA        String?               @default("")
  modelA          ModelA?         @relation("relationList", fields: [idModelA], references: [id])
  preferred   ModelB?         @relation("relationPreferred", fields: [id], references: [preferredModelBId ])

  @@index([idModelA        ])
}
```
as you can see there is a 1:N relation but on ModelA  I want to store the preferred record of model B and need a relation to show only the preferred record. 
In this case, you will have 2 relations between the table, the first will be a list and the second a 1:1 relation. You are asking to create a unique index on ModelA for preferredModelBId which is completely unnecessary. Nevertheless, I was forced to create it and now I'm not able to update the preferred id because Prisma seems not able to manage this use case.

**Reply by SevInf** [reply]
This schema is not valid: `preferred` field on `ModelB` referes to `ModelB` itself and references non-existing `preferredModelBId` field. 

If you meant to say `preferred ModelA?` instead it has the same problem and the same explanation I've already given for the previous schema you posted: without unique index on `preferredModeBId` there can be multiple `ModelA`s listing the same `ModelB` as preferred, so relation can not be 1:1.

**Reply by SevInf** [reply]
To make it more clear, suppose you have following records in the `ModelA` table:

```js
{ id: 1, prefererredModelBId: 1 }
{ id: 2, prefererredModelBId: 1 }
```

and `ModelB` table has a single record `{ id: 1}`.

What should the following query return?

```js
prisma.modelB.findUnique({
  where: { id: 1 },
  include: { preferred: true }  
})
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
